src/compiler/transformers/classFields.ts

```diff
@@ -2144,7 +2144,7 @@ namespace ts {
                 // [ { set value(x) { this.#myProp = x; } }.value ] = [ "hello" ];
                 return factory.updateArrayLiteralExpression(
                     node,
-                    visitNodes(node.elements, visitArrayAssignmentTarget, isExpression)
+                    visitNodes(node.elements as NodeArray<Extract<BindingOrAssignmentElement, Expression>>, visitArrayAssignmentTarget, isExpression)

```

Comment:
This is a case where `isExpression` does not narrow down `BindingOrAssignmentElement` to its specific components.

Maybe the most recent type guard + union changes will help?

Reply:
No, it didn't help. Sort of make it worse, even.

```
No overload matches this call.
  Overload 1 of 2, '(nodes: NodeArray<Expression>, visitor: Visitor<Expression> | undefined, test?: ((node: Expression) => boolean) | undefined, start?: number | undefined, count?: number | undefined): NodeArray<...>', gave the following error.
    Argument of type '(node: BindingOrAssignmentElement) => LeftHandSideExpression | OmittedExpression | PropertyAssignment | ... 6 more ... | SpreadElement' is not assignable to parameter of type 'Visitor<Expression>'.
      Types of parameters 'node' and 'node' are incompatible.
        Type 'Expression' is not assignable to type 'BindingOrAssignmentElement'.
          Type 'Expression' is missing the following properties from type 'AssignmentExpression<EqualsToken>': left, operatorToken, right, _declarationBrand
  Overload 2 of 2, '(nodes: NodeArray<Expression> | undefined, visitor: Visitor<Expression> | undefined, test?: ((node: Expression) => boolean) | undefined, start?: number | undefined, count?: number | undefined): NodeArray<...> | undefined', gave the following error.
    Argument of type '(node: BindingOrAssignmentElement) => LeftHandSideExpression | OmittedExpression | PropertyAssignment | ... 6 more ... | SpreadElement' is not assignable to parameter of type 'Visitor<Expression>'.ts(2769)
```

Reply:
Given that `visitNodes`'s `test` is applied to the `updated` nodes, I think there's actually a bug here; `elements` seems like it should have a narrower type than `Expression`, or that this is the `is` function to call here.